core: do not trigger rewind behaviour when modifying pre-genesis hardfork times / blocks

geoknee opened the pull request:
Replaces https://github.com/ethereum-optimism/op-geth/pull/257 

Currently, no Optimism hard forks are checked as a part of this function:
https://github.com/ethereum-optimism/op-geth/blob/0d7fd67bfa00ea6dfc2e1371ea72db001731b43d/params/config.go#L807-L876

So modifying an optimism hard fork time will not cause a chain rewind, whereas modifying a non-optimism hardfork will cause a chain rewind. 

Towards https://github.com/ethereum-optimism/client-pod/issues/918
Counterpart to https://github.com/ethereum-optimism/superchain-registry/pull/260

* adds all of the optimism hard forks to the check so they have the same behaviour
* modifies the check so that changing a hardfork from a pregenesis timestamp to another pregenesis timestamp is considered a compatible change.

geoknee committed:
skeetch out a test for config change pre genesis

geoknee committed:
expand test to multiple test cases

geoknee commented:
Next steps following sync with @sebastianst 
1. Add some unit test for `checkCompatible` and patch it to include all Optimism hard forks
2. Patch it again so that pre-genesis hardfork times are always considered compatible
3. Test out the a "test release" of op-geth on a) a devnet replica and/or b) our prod metal replica (but protect that one first by asking devinfra to take a disk snapshot prior)

geoknee commented:
> Shanghai activates with Canyon
The code that ties these together is outside of the function covered by the test on this PR. So a "live" test which say modifies a canyon time can _also_ modify the shanghai time (which would cause a chain rewind currently).

geoknee committed:
thread genesisTimestamp down to ChainConfig.CheckCompatible

geoknee committed:
add failing test cases

geoknee committed:
refactor test

geoknee committed:
add all optimism forks to checkCompatible

This is brittle in general and longer term we need a solution that requires less boiler plate.

geoknee committed:
fix: pregenesis fork changes are not incompatible

geoknee committed:
fix integration test

geoknee committed:
lint

geoknee committed:
remove genesis number from checkCompatible

it should always be 0

geoknee committed:
allow for nil genesis in SetupGenesisBlockWithOverride

geoknee commented on core/genesis.go:
```diff
@@ -376,7 +376,11 @@ func SetupGenesisBlockWithOverride(db ethdb.Database, triedb *triedb.Database, g
 		return newcfg, stored, errors.New("missing head header")
 	}
-	compatErr := storedcfg.CheckCompatible(newcfg, head.Number.Uint64(), head.Time)
+	genesisTimestamp := uint64(0)
+	if genesis != nil {
+		genesisTimestamp = genesis.Timestamp
+	}
+	compatErr := storedcfg.CheckCompatible(newcfg, head.Number.Uint64(), head.Time, genesisTimestamp)
```
This is not ideal -- an alternative might be to have `storedcfg.CheckCompatible()` take a pointer to a genesisTimestamp, and then we can pass a nil pointer in in the case where the genesis itself is nil.

sebastianst replied:
I think I like the alternative better. We already pass a uint46 ptr to the `isTimestampPreGenesis`, so we might as well also pass it to `CheckCompatible`.

sebastianst commented on core/blockchain_test.go:
```diff
@@ -4326,3 +4332,75 @@ func TestEIP3651(t *testing.T) {
 	}
 }
+
+func TestRewindOnConfigChange(t *testing.T) {
```
If we move new tests into new files like `blockchain_optimism_test.go` then we minimize the chance for merge conflicts when doing future upstream merges. This is a general pattern that we could follow, where we put new code as much as possible into new files with an `_optimism` in its name.

sebastianst commented on params/config_test.go:
```diff
@@ -27,10 +28,11 @@ import (
+		headBlock,
+		headTimestamp,
+		genesisTimestamp uint64
+		wantErr *ConfigCompatError
```
When changing existing structs in op-geth, we can minimize the actual diff by leaving the existing struct as is, then adding a new line, and only then the new fields, e.g.
```
		stored, new   *ChainConfig
		headBlock     uint64
		headTimestamp uint64
		wantErr       *ConfigCompatError

		genesisTimestamp uint64
```

geoknee replied:
[3d4cd65](https://github.com/ethereum-optimism/op-geth/pull/332/commits/3d4cd65c52dd3f4141c0c4a0dcbbf7e41d79a2b9)

sebastianst commented on params/config_test.go:
```diff
@@ -109,13 +111,45 @@ func TestCheckCompatible(t *testing.T) {
+				NewTime:      newUint64(20),
+				RewindToTime: 9,
+			},
+		},
```
what's new about this test case?

geoknee replied:
[c43b7a2](https://github.com/ethereum-optimism/op-geth/pull/332/commits/c43b7a20e79ce78b694b9ac710d9489034fe2f4c)

geoknee commented:
> Next steps following sync with @sebastianst
> 
> 1. Add some unit test for `checkCompatible` and patch it to include all Optimism hard forks

Done
> 2. Patch it again so that pre-genesis hardfork times are always considered compatible

Done
> 3. Test out the a "test release" of op-geth on a) a devnet replica and/or b) our prod metal replica (but protect that one first by asking devinfra to take a disk snapshot prior)

Done!

sebastianst commented on core/blockchain_test.go:
```diff
@@ -4326,3 +4332,75 @@ func TestEIP3651(t *testing.T) {
+
+	tcs := []testCase{
+		{
+			name:              fmt.Sprintf("CanyonTime changes from 10 to 0 (genesis time is %d)", genesisTime),
```
nit: if using this string as the test name in `t.Run`, I suggest to use a short identifier like just `canyon` here, because this whole string gets converted into the test name. In a failing scenario this would be enough to track down the failing test.

sebastianst commented on core/blockchain_test.go:
```diff
@@ -4326,3 +4332,75 @@ func TestEIP3651(t *testing.T) {
+			blockchain, _ := NewBlockChain(db, cc, genesis, nil, ethash.NewFaker(), vm.Config{}, nil, nil)
+
+			// Stop chain after 1 second
+			<-time.After(1 * time.Second)
```
There's `time.Sleep`. And is there some polling to the `blockchain` that we can do to confirm the db has been written to once, that we could poll in a loop with a shorter timeout (possibly using `require.Eventually`)? This would be quicker and at the same time less flaky than assuming that after 1sec the desired state has been achieved.

sebastianst commented on core/blockchain_test.go:
```diff
@@ -4326,3 +4332,75 @@ func TestEIP3651(t *testing.T) {
+			// Restart chain with modified genesis config
+			tc.override2(genesis.Config)
+			blockchain, _ = NewBlockChain(db, cc, genesis, nil, ethash.NewFaker(), vm.Config{}, nil, nil)
+			<-time.After(1 * time.Second)
```
same here, can we do some polling of some assertion with a shorter poll rate to make the test less flaky?

geoknee replied:
In the end I removed this test entirely as discussed (we will rely on the more reliable / focussed unit tests for this PR).

geoknee committed:
move test to its own file

geoknee committed:
use shorter test names

geoknee committed:
remove duplicate test case

geoknee committed:
reorder struct fields to reduce diff

geoknee committed:
CheckCompatible accepts a *uint64 genesisTimestamp

This allows us to handle the case where the genesis itself (and therefore the genesis timestamp) is undefined. We handle that by declaring all time stamps as POST genesis, so they can't pass the compatibilty check via the new escape hatch.

geoknee commented:
> We already pass a uint46 ptr to the isTimestampPreGenesis, so we might as well also pass it to CheckCompatible.

[2a67982](https://github.com/ethereum-optimism/op-geth/pull/332/commits/2a67982966e78b2e7ef1beff42614c5f6a967702)

geoknee committed:
remove optimism test

geoknee committed:
remove newline
